cuda_core/tests/test_program.py

```diff
@@ -9,10 +9,187 @@
+
+    _libnvvm_version_attempted = True
+
+    try:
+        from cuda.core.experimental._program import _get_nvvm_module
+
+        nvvm = _get_nvvm_module()
+
+        try:
+            from cuda.bindings.utils import get_minimal_required_cuda_ver_from_ptx_ver, get_ptx_ver
+        except ImportError:
+            _libnvvm_version = None
+            return _libnvvm_version
+
+        program = nvvm.create_program()
+        try:
+            precheck_ir_bytes = precheck_nvvm_ir.encode("utf-8")
+            nvvm.add_module_to_program(program, precheck_ir_bytes, len(precheck_ir_bytes), "precheck.ll")
+
+            options = ["-arch=compute_90"]
+            nvvm.verify_program(program, len(options), options)
+            nvvm.compile_program(program, len(options), options)
+
+            ptx_size = nvvm.get_compiled_result_size(program)
+            ptx_data = bytearray(ptx_size)
+            nvvm.get_compiled_result(program, ptx_data)
+            ptx_str = ptx_data.decode("utf-8")
+            ptx_version = get_ptx_ver(ptx_str)
+            cuda_version = get_minimal_required_cuda_ver_from_ptx_ver(ptx_version)
+            _libnvvm_version = cuda_version
+            return _libnvvm_version
+
+        finally:
+            nvvm.destroy_program(program)
+
+    except Exception:
+        _libnvvm_version = None
+        return _libnvvm_version
+
+
+@pytest.fixture(scope="session")
+def nvvm_ir():
+    """Generate working NVVM IR with proper version metadata.
+    The try clause here is used for older nvvm modules which
+    might not have an ir_version() method. In which case the
+    fallback assumes no version metadata will be present in
+    the input nvvm ir
+    """
+    try:
+        from cuda.core.experimental._program import _get_nvvm_module
+
+        nvvm = _get_nvvm_module()
+        major, minor, debug_major, debug_minor = nvvm.ir_version()
+
+        nvvm_ir_template = """target triple = "nvptx64-unknown-cuda"
+target datalayout = "e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-i128:128:128-f32:32:32-f64:64:64-v16:16:16-v32:32:32-v64:64:64-v128:128:128-n16:32:64"
+
+define i32 @ave(i32 %a, i32 %b) {{
+entry:
+  %add = add nsw i32 %a, %b
+  %div = sdiv i32 %add, 2
+  ret i32 %div
+}}
+
+define void @simple(i32* %data) {{
+entry:
+  %0 = call i32 @llvm.nvvm.read.ptx.sreg.ctaid.x()
+  %1 = call i32 @llvm.nvvm.read.ptx.sreg.ntid.x()
+  %mul = mul i32 %0, %1
+  %2 = call i32 @llvm.nvvm.read.ptx.sreg.tid.x()
+  %add = add i32 %mul, %2
+  %call = call i32 @ave(i32 %add, i32 %add)
+  %idxprom = sext i32 %add to i64
+  store i32 %call, i32* %data, align 4
+  ret void
+}}
+
+declare i32 @llvm.nvvm.read.ptx.sreg.ctaid.x() nounwind readnone
+declare i32 @llvm.nvvm.read.ptx.sreg.ntid.x() nounwind readnone
+declare i32 @llvm.nvvm.read.ptx.sreg.tid.x() nounwind readnone
+
+!nvvm.annotations = !{{!0}}
+!0 = !{{void (i32*)* @simple, !"kernel", i32 1}}
+
+!nvvmir.version = !{{!1}}
+!1 = !{{i32 {major}, i32 0, i32 {debug_major}, i32 0}}
```

Comment:
Q: For my own curiosity: `ir_version()` returns 4 integers. Why don't we use them all, instead of hard-coding the minor versions to 0?

Reply:
Yes this is modified fully.

Reply:
I am not sure I understand....? 🤔